File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.cs
using System;

public class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int gA = arrayA[0];
        int gB = arrayB[0];
        
        for(int i = 1; i < arrayA.Length; i++)
        {
            gA = GCD(gA, arrayA[i]);
        }
        
        for(int i = 1; i < arrayB.Length; i++)
        {
            gB = GCD(gB, arrayB[i]);
        }
        
        int candA = DividesNone(gA, arrayB) ? gA : 0;
        int candB = DividesNone(gB, arrayA) ? gB : 0;
        
        return Math.Max(candA, candB);
    }
    
    private int GCD(int a, int b)
    {
        while(b != 0)
        {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
    
    private bool DividesNone(int d, int[] arr)
    {
        foreach(var x in arr)
        {
            if(x % d == 0) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int go(vector<int> &array)
{
    int g = array[0];
    for(int i = 1; i < array.size(); i++)
    {
        g = gcd(g, array[i]);
    }
    return g;
}

bool check(int g, vector<int> &array)
{
    for(int num : array)
    {
        if(num % g == 0)
            return false;
    }
    return true;
}

int solution(vector<int> arrayA, vector<int> arrayB) {
    int answer = 0;
    
    int gcdA = go(arrayA);
    int gcdB = go(arrayB);
    
    if(check(gcdA, arrayB))
        answer = max(answer, gcdA);
    if(check(gcdB, arrayA))
        answer = max(answer, gcdB);

    return answer;
}